### PR TITLE
Introduce mypy and enforce it on a few files

### PR DESCRIPTION
**Pull request by asottile**
to be honest, I'm not exactly sure how I got here -- I was reading through the project to try and understand what it does and noticed #15 and somehow my first thought was "oh hey, that sounds like fun".  I applied it to a few modules and then got to a point where I needed to actually understand the code to make further progress and so I decided to commit what I had and make a PR.

This also (selfishly?) adds [pre-commit](https://pre-commit.com) as an easier way to run linting (~I noticed there wasn't currently any automated linting in CI so I added it!~ EDIT: had to change from pep8 -> flake8 so that type annotations weren't false positives) -- this part isn't entirely necessary as mypy can be invoked directly.

### CODE REVIEW

File: src/bandersnatch/configuration.py [hunk 1]
```diff
@@ -1,14 +1,19 @@
 """
 Module containing classes to access the bandersnatch configuration file
 """
+from typing import Any
+from typing import Dict
+from typing import Optional
+from typing import Type
```
Comment (cooperlees) [inline src/bandersnatch/configuration.py]: Why not all one from?
i.e.
from typing import (Any, Dict, Optional, Type)

File: src/bandersnatch/utils.py [hunk 2]
```diff
@@ -89,7 +94,7 @@ def update_safe(filename, **kw):
             prefix=os.path.basename(filename) + '.', **kw) as tf:
         if os.path.exists(filename):
             os.chmod(tf.name, os.stat(filename).st_mode & 0o7777)
-        tf.has_changed = False
+        tf.has_changed = False  # type: ignore
```
Comment (cooperlees) [inline src/bandersnatch/utils.py]: What's the error here? I feel we should be able to fix it.

File: test-requirements.txt [hunk 3]
```diff
@@ -1,7 +1,4 @@
-flake8
```
Comment (cooperlees) [inline test-requirements.txt]: Don't we still use flake8?

File: src/bandersnatch/configuration.py (hunk 1 above)
Comment (asottile) [reply src/bandersnatch/configuration.py]: personal preference though if you'd like a different style I will change it.

I use [reorder_python_imports](https://github.com/asottile/reorder_python_imports) -- the style here aims to reduce merge conflicts above everything else.

If you'd like, I could set you up with `isort` in this PR as well, that way we don't have to bicker about needless style

File: src/bandersnatch/utils.py (hunk 2 above)
Comment (asottile) [reply src/bandersnatch/utils.py]: `has_changed` is not an attribute of file objects, there's attribute stuffing happening here

File: test-requirements.txt (hunk 3 above)
Comment (asottile) [reply test-requirements.txt]: yes, but via pre-commit.  pre-commit manages the installation in a separate environment so you don't have to worry about system installation of tools

**Comment by cooperlees** [comment]
All sounds good - If you're down to add `isort` too that would be awesome. Then I'll follow up with a black style pre-commit + PR so style is behind us :)

File: test-requirements.txt (hunk 3 above)
Comment (cooperlees) [reply test-requirements.txt]: I only like people to have the ability to run things manually.
- Sorry about all the questions, I am a GitHub noob, been locked away in Facebook for to long.

I'm not quite Dave Harrington n00b, but close.

File: tox.ini [hunk 4]
```diff
@@ -19,3 +19,14 @@ extras = doc_build
...
+
+[testenv:lint]
+basepython=python3
+skip_install=True
+deps=pre-commit
+commands=
+    pre-commit run --all-files --show-diff-on-failure
+
+[isort]
+include_trailing_comma = True
```
Comment (cooperlees) [inline tox.ini]: I'll fix this if black does not agree here. I can't remember.

**Comment by cooperlees** [comment]
I'll also plan to use MonkeyType or like tool to get more type annotations for free.

File: test-requirements.txt (hunk 3 above)
Comment (asottile) [reply test-requirements.txt]: oh hey there @DaveHarrington -- small world I guess?

**Pull request merged**